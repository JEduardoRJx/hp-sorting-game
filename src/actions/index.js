export const setCharacters = characters => ({
  type: 'SET_CHAR',
  characters
})

export const setGameCharacters = allCharacters => ({
  type: 'SET_GAME_CHAR',
  allCharacters
})

export const isLoading = bool => ({
  type: 'IS_LOADING',
  bool
})

export const setHouse = house => ({
  type: 'SET_HOUSE',
  house
})