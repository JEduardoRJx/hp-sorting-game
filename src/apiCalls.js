const baseUrl = 'https://www.potterapi.com/v1';

const apiKey = `?key=$2a$10$soGCoAQkdenPDKuyeTco3urLugvc4a/RMC2S3tzviQxAbVxmHAPN6`

const apiKey2 = `?key=$2a$10$AZk7aQ1VZuDe/U1v8A.9u.k/U7knzlj8B7gonTU6Qw/mtqV4rO6zq`

export const fetchCharacters = async () => {
  const url = `${baseUrl}/characters${apiKey2}`;
  const response = await fetch(url);
  if(!response.ok) {
    throw Error('There\'s an error with your wand, can\'t fetch characters');
  }
  const characters = await response.json();
  return characters;
}

export const fetchSortingHat = async () => {
  const url = `${baseUrl}/sortingHat`;
  const response = await fetch(url);
  if(!response.ok) {
    throw Error('The sorting hat must me asleep, can\'t fetch House');
  }
  const house = await response.json();
  return house;
}