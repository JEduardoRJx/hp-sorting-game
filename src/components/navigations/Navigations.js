import React from 'react';
import { NavLink } from 'react-router-dom';
import './Navigations.scss';

const Navigations = () => {
  return (
    <div className='navigation'>
      <div className='icon-wrapper sorting-hat'>
        <NavLink to='/discoveryourhouse'>
          <img className='main-screen-img' 
            src={require('../../images/sortingHat.svg')} 
            alt='sorting hat' />
        </NavLink>
        <label>Discover Your House</label>
      </div>
      <div className='icon-wrapper golden-snitch'>
        <NavLink to='/game'>
          <img className='main-screen-img' 
            src={require('../../images/snitch.svg')} 
            alt='golden snitch' />
        </NavLink>
        <label>Play Game</label>
      </div>
      <div className='icon-wrapper instructions'>
        <NavLink to='/instructions'>
          <img className='main-screen-img' 
            src={require('../../images/instructions.svg')} 
            alt='instructions' />
        </NavLink>
        <label>Instructions</label>
      </div>
      <div className='icon-wrapper characters'>
        <NavLink to='/characters'>
          <img className='main-screen-img' 
            src={require('../../images/characters.svg')} 
            alt='characters' />
        </NavLink>
        <label>View All Characters</label>
      </div>
      <div className='icon-wrapper wand'>
        <NavLink to='/spells'>
          <img className='main-screen-img' 
            src={require('../../images/wand.svg')} 
            alt='wand view spells' />
        </NavLink>
        <label>View All Spells</label>
      </div>
    </div>
  )
}

export default Navigations