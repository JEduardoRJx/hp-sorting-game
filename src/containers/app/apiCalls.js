const baseUrl = 'https://www.potterapi.com/v1';
const sortingHat = '/sortingHat';
const houses = '/houses'
const spells = '/spells'

const apiKey = `?key=$2a$10$soGCoAQkdenPDKuyeTco3urLugvc4a/RMC2S3tzviQxAbVxmHAPN6`

export const fetchCharacters = async () => {
  const url = `${baseUrl}/characters${apiKey}`;
  const response = await fetch(url);
  if(!response.ok) {
    throw Error('There\'s an error with your wand, can\'t fetch characters');
  }
  const characters = await response.json();
  return characters;
}