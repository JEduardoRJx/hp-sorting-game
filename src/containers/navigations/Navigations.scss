$hw-dark: #4c3638;
$hw-light: #D3A481;
$sl-dark: #276F45;
$sl-light: #A8B0BF;
$gryf-dark: #702737;
$gryf-light: #DAA520;
$rav-dark: #274F70;
$rav-light: #C98C4E;
$huff-dark: #141414;
$huff-light: #D1B738;

.navigation {
  grid-column: 2/4;
  width: 100%;
  display: grid;
  grid-template-columns: 1fr 1fr;
}

.icon-wrapper {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.main-screen-img {
  display: flex;
  height: 185px;
  &:hover {
    transform: scale(1.1);
    cursor: pointer;
  }
}

.sorting-hat {
  grid-column: 1/3;
  // border-bottom: 5px solid $hw-dark;
}

label {
  font-size: 1.5em;
  font-weight: bold;
  margin-top: 10px;
}

.sorting-hat,
.characters,
.instructions {
  background-color: $hw-light;
  color: $hw-dark;
  &.gryffindor {
    background-color: $gryf-light;
    color: $gryf-dark;
  }
  &.slytherin {
    background-color: $sl-light;
    color: $sl-dark;
  }
  &.hufflepuff {
    background-color: $huff-light;
    color: $huff-dark;
  }
  &.ravenclaw {
    background-color: $rav-light;
    color: $rav-dark;
  }
}

.golden-snitch,
.wand {
  color: $hw-light;
  &.gryffindor {
    color: $gryf-light;
  }
  &.slytherin {
    color: $sl-light;
  }
  &.hufflepuff {
    color: $huff-light;
  }
  &.ravenclaw {
    color: $rav-light;
  }
}