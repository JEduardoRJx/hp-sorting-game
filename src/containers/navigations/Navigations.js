import React, { Component } from 'react';
import { NavLink } from 'react-router-dom';
import './Navigations.scss';
import { connect } from 'react-redux';
import { setGameCharacters, setHouse, isLoading } from '../../actions';
import { fetchSortingHat } from '../../apiCalls';

export class Navigations extends Component {

  selectHouse = async () => {
    const { setHouse, loading } = this.props
    try {
      loading(true);
      const house = await fetchSortingHat()
      setHouse(house)
      loading(false)
    } catch ({ message }) {
      console.log(message)
    }
  }

  render() {
    const { setGameCharacters, allCharacters, user} = this.props
    const house = user === undefined ? '' : user.house
    return (
      <div className='navigation'>
        <div className={`icon-wrapper sorting-hat ${house}`}>
          <NavLink to='/discoveryourhouse' 
            onClick={this.selectHouse}>
            <img className='main-screen-img' 
              src={require('../../images/sortingHat.svg')} 
              alt='sorting hat' />
          </NavLink>
          <label>Discover Your House</label>
        </div>
        <div className={`icon-wrapper golden-snitch ${house}`}>
          <NavLink to='/game' onClick={() => setGameCharacters(allCharacters)}>
            <img className='main-screen-img' 
              src={require('../../images/snitch.svg')} 
              alt='golden snitch' />
          </NavLink>
          <label>Play Game</label>
        </div>
        <div className={`icon-wrapper instructions ${house}`}>
          <NavLink to='/instructions'>
            <img className='main-screen-img' 
              src={require('../../images/instructions.svg')} 
              alt='instructions' />
          </NavLink>
          <label>Instructions</label>
        </div>
        <div className={`icon-wrapper characters ${house}`}>
          <NavLink to='/characters'>
            <img className='main-screen-img' 
              src={require('../../images/characters.svg')} 
              alt='characters' />
          </NavLink>
          <label>View All Characters</label>
        </div>
        <div className={`icon-wrapper wand ${house}`}>
          <NavLink to='/spells'>
            <img className='main-screen-img' 
              src={require('../../images/wand.svg')} 
              alt='wand view spells' />
          </NavLink>
          <label>View All Spells</label>
        </div>
      </div>
    )
  }
}

export const mapStateToProps = ({ allCharacters, user }) => ({
  allCharacters,
  user
})

export const mapDispatchToProps = dispatch => ({
  setGameCharacters: characters => dispatch( setGameCharacters(characters) ),
  loading: bool => dispatch( isLoading(bool) ),
  setHouse: house => dispatch( setHouse(house) )
})

export default connect(mapStateToProps, mapDispatchToProps)(Navigations)