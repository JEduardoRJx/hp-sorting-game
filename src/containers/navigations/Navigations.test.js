import React from 'react';
import { shallow } from 'enzyme';
import { Navigations, mapStateToProps, mapDispatchToProps } from './Navigations';
import { fetchSortingHat } from '../../apiCalls'
jest.mock('../../apiCalls');
import { setGameCharacters, setHouse, isLoading } from '../../actions';

describe('NavigationsContainer', () => {

  describe('Navigations Component', () => {
    let wrapper;
    const mockSetGameCharacters = jest.fn();
    const mockLoading = jest.fn();
    const mockSetHouse = jest.fn();
    const mockAllCharacters = [{
      bloodStatus: "pure-blood",
      deathEater: false,
      dumbledoresArmy: false,
      house: "Hufflepuff",
      ministryOfMagic: false,
      name: "Cedric Diggory",
      orderOfThePhoenix: false,
      role: "Student",
      school: "Hogwarts School of Witchcraft, and Wizardry",
      species: "human",
      wand: 'Ash, 12 1/4", unicorn hair',
      __v: 0,
      _id: "5a1096253dc2080021cd875f"
    }]
    const mockUser = {
      name: 'name',
      wizardkind: 'wizard',
      wand: 'wand',
      house: 'slytherin'
    }
    fetchSortingHat.mockImplementation(() => Promise.resolve('slytherin'))

    beforeEach(() => {
      wrapper = shallow (<Navigations 
        allCharacters={mockAllCharacters}
        user={mockUser}
        setGameCharacters={mockSetGameCharacters}
        loading={mockLoading}
        setHouse={mockSetHouse}
      />);
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should setHouse when selectHouse is called', () => {
      wrapper.instance().selectHouse();
      expect(fetchSortingHat).toHaveBeenCalled();
      expect(mockLoading).toHaveBeenCalledWith(true);
    });

    it('should call setGameCharacters on click', () => {
      wrapper.find('NavLink').at(1).simulate('click');
      expect(mockSetGameCharacters).toHaveBeenCalled();
    })
  });

  describe('mapStateToProps', () => {
    
    it('should return an allCharacters array', () => {
      const mockState = {
        filter: 'SET_CHAR',
        allCharacters: [{
          bloodStatus: "pure-blood",
          deathEater: false,
          dumbledoresArmy: false,
          house: "Hufflepuff",
          ministryOfMagic: false,
          name: "Cedric Diggory",
          orderOfThePhoenix: false,
          role: "Student",
          school: "Hogwarts School of Witchcraft, and Wizardry",
          species: "human",
          wand: 'Ash, 12 1/4", unicorn hair',
          __v: 0,
          _id: "5a1096253dc2080021cd875f"
        }]
      }
      const expected = {
        allCharacters: [{
          bloodStatus: "pure-blood",
          deathEater: false,
          dumbledoresArmy: false,
          house: "Hufflepuff",
          ministryOfMagic: false,
          name: "Cedric Diggory",
          orderOfThePhoenix: false,
          role: "Student",
          school: "Hogwarts School of Witchcraft, and Wizardry",
          species: "human",
          wand: 'Ash, 12 1/4", unicorn hair',
          __v: 0,
          _id: "5a1096253dc2080021cd875f"
        }]
      }
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
  });

  it('should return user', () => {
    const mockState = {
      filter: 'SET_USER',
      user: {
        name: 'name',
        wizardkind: 'wizard',
        wand: 'wand',
        house: 'slytherin'
      }
    }
    const expected = {
      user: {
        name: 'name',
        wizardkind: 'wizard',
        wand: 'wand',
        house: 'slytherin'
      }
    }
    const mappedProps = mapStateToProps(mockState);
    expect(mappedProps).toEqual(expected)
  });
});


  describe('mapDispatchToProps', () => {
    
    it('calls dispatch with characters when setGameCharacters is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = setGameCharacters([{
        bloodStatus: "pure-blood",
        deathEater: false,
        dumbledoresArmy: false,
        house: "Hufflepuff",
        ministryOfMagic: false,
        name: "Cedric Diggory",
        orderOfThePhoenix: false,
        role: "Student",
        school: "Hogwarts School of Witchcraft, and Wizardry",
        species: "human",
        wand: 'Ash, 12 1/4", unicorn hair',
        __v: 0,
        _id: "5a1096253dc2080021cd875f"
      }]);

      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.setGameCharacters([{
        bloodStatus: "pure-blood",
        deathEater: false,
        dumbledoresArmy: false,
        house: "Hufflepuff",
        ministryOfMagic: false,
        name: "Cedric Diggory",
        orderOfThePhoenix: false,
        role: "Student",
        school: "Hogwarts School of Witchcraft, and Wizardry",
        species: "human",
        wand: 'Ash, 12 1/4", unicorn hair',
        __v: 0,
        _id: "5a1096253dc2080021cd875f"
      }]);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('calls dispatch with boolean when isLoading is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = isLoading(true);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.loading(true);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('calls dispatch with a house when setHouse is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = setHouse('slytherin');
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setHouse('slytherin');
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});