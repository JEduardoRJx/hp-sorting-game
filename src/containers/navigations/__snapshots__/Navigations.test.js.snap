// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationsContainer Navigations Component should match snapshot 1`] = `
<div
  className="navigation"
>
  <div
    className="icon-wrapper sorting-hat slytherin"
  >
    <NavLink
      onClick={[Function]}
      to="/discoveryourhouse"
    >
      <img
        alt="sorting hat"
        className="main-screen-img"
        src={
          Object {
            "ReactComponent": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "__esModule": true,
            "default": "sortingHat.svg",
          }
        }
      />
    </NavLink>
    <label>
      Discover Your House
    </label>
  </div>
  <div
    className="icon-wrapper golden-snitch slytherin"
  >
    <NavLink
      onClick={[Function]}
      to="/game"
    >
      <img
        alt="golden snitch"
        className="main-screen-img"
        src={
          Object {
            "ReactComponent": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "__esModule": true,
            "default": "snitch.svg",
          }
        }
      />
    </NavLink>
    <label>
      Play Game
    </label>
  </div>
  <div
    className="icon-wrapper instructions slytherin"
  >
    <NavLink
      to="/instructions"
    >
      <img
        alt="instructions"
        className="main-screen-img"
        src={
          Object {
            "ReactComponent": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "__esModule": true,
            "default": "instructions.svg",
          }
        }
      />
    </NavLink>
    <label>
      Instructions
    </label>
  </div>
  <div
    className="icon-wrapper characters slytherin"
  >
    <NavLink
      to="/characters"
    >
      <img
        alt="characters"
        className="main-screen-img"
        src={
          Object {
            "ReactComponent": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "__esModule": true,
            "default": "characters.svg",
          }
        }
      />
    </NavLink>
    <label>
      View All Characters
    </label>
  </div>
  <div
    className="icon-wrapper wand slytherin"
  >
    <NavLink
      to="/spells"
    >
      <img
        alt="wand view spells"
        className="main-screen-img"
        src={
          Object {
            "ReactComponent": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "__esModule": true,
            "default": "wand.svg",
          }
        }
      />
    </NavLink>
    <label>
      View All Spells
    </label>
  </div>
</div>
`;
